cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ConnectorDag.java

```diff
@@ -0,0 +1,287 @@
+        and break it down into separate phases. For example:
+
+                    |---> reduce1 ---|
+          source ---|                |---> sink
+                    |---> reduce2 ---|
+
+        From the previous section, both reduces will get a connector inserted in front:
+
+                    |---> reduce1.connector               reduce1.connector ---> reduce1 ---|
+          source ---|                              =>                                       |---> sink
+                    |---> reduce2.connector               reduce2.connector ---> reduce2 ---|
+
+        Since we don't support multi-input yet, we need to convert that further into 3 phases:
+
+          connector1 ---> reduce1 ---> sink.connector
+                                                            =>       sink.connector ---> sink
+          connector2 ---> reduce2 ---> sink.connector
+
+        To find these nodes, we traverse the graph in order and keep track of sources that have a path to each node
+        with a map of node -> [ sources that have a path to the node ]
+        if we find that a node is accessible by more than one source, we insert a connector in front of it and
+        reset all sources for that node to its connector
+     */
+    SetMultimap<String, String> nodeSources = HashMultimap.create();
+    for (String source : sources) {
+      nodeSources.put(source, source);
+    }
+    for (String connector : connectors) {
+      nodeSources.put(connector, connector);
+    }
+    for (String node : linearize()) {
+      if (sinks.contains(node)) {
+        continue;
+      }
+      Set<String> connectedSources = nodeSources.get(node);
+      // if more than one source is connected to this node, then this node is a connector.
+      // keep track of it, and wipe its node sources to just contain itself
+      if (connectedSources.size() > 1) {
+        addConnectorInFrontOf(node, addedAlready);
+        connectedSources = new HashSet<>();
+        connectedSources.add(node);
+        nodeSources.replaceValues(node, connectedSources);
+      }
+      for (String nodeOutput : getNodeOutputs(node)) {
+        // don't propagate to connectors, since they are counted as source and replace any inputs they may have
+        if (connectors.contains(nodeOutput)) {
+          continue;
+        }
+        // propagate sources connected to me to all my outputs
+        nodeSources.putAll(nodeOutput, connectedSources);
+      }
+    }
+
+    /*
+        Find reduce nodes that are accessible from other reduce nodes. For example:
+
+          source ---> reduce1 ---> reduce2 ---> sink
+
+        Needs to be broken down into:
+
+          source ---> reduce1 ---> reduce2.connector      =>     reduce2.connector ---> reduce2 ---> sink
+     */
+    for (String reduceNode : reduceNodes) {
+      Set<String> accessibleByNode = accessibleFrom(reduceNode, Sets.union(connectors, reduceNodes));
```

Comment:
the 2nd argument to accessibleFrom tells the method to stop searching when it sees any node in that set.  It effectively cuts off the graph at those nodes for the purposes of that search.